File: Data Structures & Algorithms/depth-of-binary-tree/submission-0.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func maxDepth(root *TreeNode) int {
    if root == nil {
        return 0
    }
    maxLeft := calcMaxDepth(root.Left, 1)
    maxRight := calcMaxDepth(root.Right, 1)
    return max(maxLeft, maxRight)+1
}

func calcMaxDepth(root *TreeNode, depth int) int {
    if root == nil {
        return depth -1
    }
    maxLeft := calcMaxDepth(root.Left, depth + 1)
    maxRight := calcMaxDepth(root.Right, depth + 1)
    return max(maxLeft, maxRight)
}

func max(a,b int) int {
    if a > b {
        return a
    }
    return b
}

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        def findKthSmallest(root: Optional[TreeNode], k: int) -> tuple[int, int]:
            if root is None:
                return (0,k)
            res, tmp = findKthSmallest(root.left, k)
            if tmp==0:
                return res, tmp
            k = tmp
            if k - 1 == 0:
                return root.val, 0

            return findKthSmallest(root.right, k-1)
        return findKthSmallest(root, k)[0]


             